server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/PointsSortedDocsProducer.java

```diff
@@ -0,0 +1,181 @@
+        super(field);
+        this.bucketFunction = bucketFunction;
+        this.lowerPointQuery = lowerPointQuery;
+        this.upperPointQuery = upperPointQuery;
+    }
+
+    @Override
+    DocIdSet processLeaf(Query query, CompositeValuesCollectorQueue queue,
+                         LeafReaderContext context, boolean fillDocIdSet) throws IOException {
+        final PointValues values = context.reader().getPointValues(field);
+        if (values == null) {
+            // no value for the field
+            return DocIdSet.EMPTY;
+        }
+        long lowerBucket = Long.MIN_VALUE;
+        Comparable<?> lowerValue = queue.getLowerValueLeadSource();
+        if (lowerValue != null) {
+            if (lowerValue.getClass() != Long.class) {
+                throw new IllegalStateException("expected Long, got " + lowerValue.getClass());
+            }
+            lowerBucket = (Long) lowerValue;
+        }
+
+        long upperBucket = Long.MAX_VALUE;
+        Comparable<?> upperValue = queue.getUpperValueLeadSource();
+        if (upperValue != null) {
+            if (upperValue.getClass() != Long.class) {
+                throw new IllegalStateException("expected Long, got " + upperValue.getClass());
+            }
+            upperBucket = (Long) upperValue;
+        }
+        DocIdSetBuilder builder = fillDocIdSet ? new DocIdSetBuilder(context.reader().maxDoc(), values, field) : null;
+        Visitor visitor = new Visitor(context, queue, builder, values.getBytesPerDimension(), lowerBucket, upperBucket);
+        try {
+            values.intersect(visitor);
+            visitor.flush();
+        } catch (CollectionTerminatedException exc) {}
+        return fillDocIdSet ? builder.build() : DocIdSet.EMPTY;
+    }
+
+    private class Visitor implements PointValues.IntersectVisitor {
+        final LeafReaderContext context;
+        final CompositeValuesCollectorQueue queue;
+        final DocIdSetBuilder builder;
+        final int maxDoc;
+        final int bytesPerDim;
+        final long lowerBucket;
+        final long upperBucket;
+
+        DocIdSetBuilder bucketDocsBuilder;
+        DocIdSetBuilder.BulkAdder adder;
+        int remaining;
+        long lastBucket;
+        boolean first = true;
+
+        Visitor(LeafReaderContext context, CompositeValuesCollectorQueue queue, DocIdSetBuilder builder,
+                int bytesPerDim, long lowerBucket, long upperBucket) {
+            this.context = context;
+            this.maxDoc = context.reader().maxDoc();
+            this.queue = queue;
+            this.builder = builder;
+            this.lowerBucket = lowerBucket;
+            this.upperBucket = upperBucket;
+            this.bucketDocsBuilder = new DocIdSetBuilder(maxDoc);
+            this.bytesPerDim = bytesPerDim;
+        }
+
+        @Override
+        public void grow(int count) {
+            remaining = count;
```

Comment:
I don't think you need to count the number of remaining docs here, the BKD tree does it for you

Reply:
I need it because we build one doc id set per bucket (not per bkd leaf) so if the values are different inside a leaf I need to know the number of remaining docs in that leaf to create the new doc id set builder.

Reply:
I see, thanks for explaining.